src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -573,11 +707,27 @@ public static void SetCreationTime(string fullPath, DateTimeOffset time, bool as
             info.CreationTimeCore = time;
         }
 
+        public static void SetCreationTime(SafeFileHandle fileHandle, DateTimeOffset time)
+        {
+            ThrowHelper.ThrowForMissingPath_SafeFileHandle(fileHandle.Path);
+            SetCreationTime(fileHandle.Path!, time, false);
```

Comment:
This will change the link creation time, and not the target.

We musn't use `FSOPT_NOFOLLOW` when calling `setattrlist` on OSX.

Because of the differences between OSX and Linux, I think it is best to move `SetCreationTime(SafeFileHandle fileHandle, DateTimeOffset time)` into separate implementations in `FileStatus.SetTimes.{OSX,OtherUnix}.cs`.

In the OSX file you can add a `bool followLinks` to `SetCreationTimeCore` that defaults to `false`. When it is set to `true`, don't add the `FSOPT_NOFOLLOW`.

The `SafeHandle` method can call it with the arg set to `true`.

Reply:
The problem with Unix changes is that it's hard to debug, since my main development system is Windows.

Would it be a big effort for you to help me with this change?

Reply:
I'll look into it, and make a commit you can add to your branch.

Reply:
We are duplicating code here from `FileStatus` in `FileSystem`.
It would be nice if all of this logic could remain in `FileStatus`, and we try limiting the duplication.
https://github.com/dotnet/runtime/pull/62721 would help to do this.

@Jozkee @carlossanlop @danmoseley wdyt?

For example, for `SetAttributes` this means:

```diff
diff --git a/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs b/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
index 5afbb39cc60..11f4864d547 100644
--- a/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
@@ -627,60 +627,7 @@ public static void SetAttributes(string fullPath, FileAttributes attributes)
         }
 
         public static void SetAttributes(SafeFileHandle fileHandle, FileAttributes attributes)
-        {
-            int result = Interop.Sys.FStat(fileHandle, out Interop.Sys.FileStatus output);
-            if (result != 0)
-            {
-                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(result), fileHandle.Path);
-            }
-
-            // Validate that only flags from the attribute are being provided.  This is an
-                // approximation for the validation done by the Win32 function.
-                const FileAttributes allValidFlags =
-                FileAttributes.Archive | FileAttributes.Compressed | FileAttributes.Device |
-                FileAttributes.Directory | FileAttributes.Encrypted | FileAttributes.Hidden |
-                FileAttributes.IntegrityStream | FileAttributes.Normal | FileAttributes.NoScrubData |
-                FileAttributes.NotContentIndexed | FileAttributes.Offline | FileAttributes.ReadOnly |
-                FileAttributes.ReparsePoint | FileAttributes.SparseFile | FileAttributes.System |
-                FileAttributes.Temporary;
-            if ((attributes & ~allValidFlags) != 0)
-            {
-                // Using constant string for argument to match historical throw
-                throw new ArgumentException(SR.Arg_InvalidFileAttrs, "Attributes");
-            }
-
-            if (Interop.Sys.CanSetHiddenFlag)
-            {
-                if ((attributes & FileAttributes.Hidden) != 0 && (output.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == 0)
-                {
-                    Interop.Sys.FChflags(fileHandle, (output.UserFlags | (uint)Interop.Sys.UserFlags.UF_HIDDEN));
-                }
-            }
-
-            // The only thing we can reasonably change is whether the file object is readonly by changing permissions.
-
-            int newMode = output.Mode;
-            if ((attributes & FileAttributes.ReadOnly) != 0)
-            {
-                // Take away all write permissions from user/group/everyone
-                newMode &= ~(int)(Interop.Sys.Permissions.S_IWUSR | Interop.Sys.Permissions.S_IWGRP | Interop.Sys.Permissions.S_IWOTH);
-            }
-            else if ((newMode & (int)Interop.Sys.Permissions.S_IRUSR) != 0)
-            {
-                // Give write permission to the owner if the owner has read permission
-                newMode |= (int)Interop.Sys.Permissions.S_IWUSR;
-            }
-
-            // Change the permissions on the file
-            if (newMode != output.Mode)
-            {
-                int chModResult = Interop.Sys.FChMod(fileHandle, newMode);
-                if (chModResult != 0)
-                {
-                    throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(chModResult), fileHandle.Path);
-                }
-            }
-        }
+            => default(FileStatus).SetAttributes(fileHandle, attributes);
 
         public static DateTimeOffset GetCreationTime(string fullPath)
         {

diff --git a/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs b/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs
index 05b4bbcff6a..9f082059c46 100644
--- a/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs
+++ b/src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs
@@ -197,6 +197,12 @@ internal FileAttributes GetAttributes(ReadOnlySpan<char> path, ReadOnlySpan<char
         }
 
         internal void SetAttributes(string path, FileAttributes attributes)
+            => SetAttributes(handle: null, path, attributes);
+
+        internal void SetAttributes(SafeFileHandle handle, FileAttributes attributes)
+            => SetAttributes(handle, handle.Path, attributes);
+
+        internal void SetAttributes(SafeFileHandle? handle, string path, FileAttributes attributes)
         {
             // Validate that only flags from the attribute are being provided.  This is an
             // approximation for the validation done by the Win32 function.
@@ -213,7 +219,7 @@ internal void SetAttributes(string path, FileAttributes attributes)
                 throw new ArgumentException(SR.Arg_InvalidFileAttrs, "Attributes");
             }
 
-            EnsureCachesInitialized(path);
+            EnsureCachesInitialized(handle, path);
 
             if (!_exists)
                 FileSystemInfo.ThrowNotFound(path);
@@ -222,8 +228,15 @@ internal void SetAttributes(string path, FileAttributes attributes)
             {
                 if ((attributes & FileAttributes.Hidden) != 0 && (_fileCache.UserFlags & (uint)Interop.Sys.UserFlags.UF_HIDDEN) == 0)
                 {
-                    // If Hidden flag is set and cached file status does not have the flag set then set it
-                    Interop.CheckIo(Interop.Sys.LChflags(path, (_fileCache.UserFlags | (uint)Interop.Sys.UserFlags.UF_HIDDEN)), path, InitiallyDirectory);
+                    if (handle is not null)
+                    {
+                        Interop.CheckIo(Interop.Sys.FChflags(handle, (output.UserFlags | (uint)Interop.Sys.UserFlags.UF_HIDDEN)), path, InitiallyDirectory);
+                    }
+                    else
+                    {
+                        // If Hidden flag is set and cached file status does not have the flag set then set it
+                        Interop.CheckIo(Interop.Sys.LChflags(path, (_fileCache.UserFlags | (uint)Interop.Sys.UserFlags.UF_HIDDEN)), path, InitiallyDirectory);
+                    }
                 }
                 else if (HasHiddenFlag)
                 {
@@ -249,7 +262,14 @@ internal void SetAttributes(string path, FileAttributes attributes)
             // Change the permissions on the file
             if (newMode != _fileCache.Mode)
             {
-                Interop.CheckIo(Interop.Sys.ChMod(path, newMode), path, InitiallyDirectory);
+                if (handle is not null)
+                {
+                    Interop.CheckIo(Interop.Sys.FChMod(handle, newMode), path, InitiallyDirectory);
+                }
+                else
+                {
+                    Interop.CheckIo(Interop.Sys.ChMod(path, newMode), path, InitiallyDirectory);
+                }
             }
 
             _initializedFileCache = -1;
@@ -384,29 +404,39 @@ internal long GetLength(ReadOnlySpan<char> path, bool continueOnError = false)
 
         // Tries to refresh the lstat cache (_fileCache).
         // This method should not throw. Instead, we store the results, and we will throw when the user attempts to access any of the properties when there was a failure
-        internal void RefreshCaches(ReadOnlySpan<char> path)
+        internal void RefreshCaches(SafeFileHandle? handle, ReadOnlySpan<char> path)
         {
             _isDirectory = false;

+            if (handle is not null)
             {
-                _exists = false;
-                return;
+                // ...
+                // Interop.Sys.FStat(fileHandle, out Interop.Sys.FileStatus output);
+                // ...
             }
+            else
+            {
+                path = Path.TrimEndingDirectorySeparator(path);
+                // ...
             }
 
             _exists = true;
@@ -420,11 +450,11 @@ internal void RefreshCaches(ReadOnlySpan<char> path)
 
         // Checks if the file cache is set to -1 and refreshes it's value.
         // If it failed, and continueOnError is set to true, this method will throw.
-        internal void EnsureCachesInitialized(ReadOnlySpan<char> path, bool continueOnError = false)
+        internal void EnsureCachesInitialized(SafeFileHandle? handle, ReadOnlySpan<char> path, bool continueOnError = false)
         {
             if (_initializedFileCache == -1)
             {
-                RefreshCaches(path);
+                RefreshCaches(handle, path);
             }
 
             if (!continueOnError)
```

Reply:
> I'll look into it, and make a commit you can add to your branch.

Thank you :smile:

Reply:
> The problem with Unix changes is that it's hard to debug, since my main development system is Windows.

@deeprobin we all use WSL2 in this situation and (being essentially a virtual machine) it has remarkably good fidelity with a "real" Linux system. You can connect to it with Visual Studio Code running on Windows - it makes it amazingly easy. With a few settings adjusted, you can step through code as well, but personally on Linux I mostly just pull down the branch, build and run tests to validate things.

Reply:
@deeprobin I plan to look at this Monday. If you want to try tackle this using WSL, let me know, and I'll wait.

Reply:
@tmds All right. I'll have a look at WSL in general at some point later, but I'm currently very busy, so I'd be happy if you take a look at it :D

Reply:
@deeprobin can you merge https://github.com/tmds/runtime/tree/safehandle_filestatus into your branch?

There are a few test failures to be tackled still. It would be good to have a cross-platform CI run before tackling them.

Reply:
@tmds I guess I will look into that during the course of tomorrow or the day after. 
Thank you very much for your effort.